main/src/ca/uwaterloo/flix/language/phase/monomorph/Lowering.scala

```diff
@@ -10,8 +43,80 @@ object Lowering {
     * When implemented:
     * Replaces schema types with the Datalog enum type and channel-related types with the channel enum type.
     *
-    * @param t the type to be lowered.
+    * @param tpe0 the type to be lowered.
     * @return
     */
-   def lowerType(t: Type): Type = t
+   def lowerType(tpe0: Type): Type = tpe0 match {
+     case Type.Cst(_, _) => tpe0 // Performance: Reuse tpe0.
+
+     case Type.Var(_, _) => tpe0
+
+     // Rewrite Sender[t] to Concurrent.Channel.Mpmc[t, IO]
+     case Type.Apply(Type.Cst(TypeConstructor.Sender, loc), tpe, _) =>
+       val t = lowerType(tpe)
+       mkChannelTpe(t, loc)
+
+     // Rewrite Receiver[t] to Concurrent.Channel.Mpmc[t, IO]
+     case Type.Apply(Type.Cst(TypeConstructor.Receiver, loc), tpe, _) =>
+       val t = lowerType(tpe)
+       mkChannelTpe(t, loc)
+
+     case Type.Apply(tpe1, tpe2, loc) =>
+       val t1 = lowerType(tpe1)
+       val t2 = lowerType(tpe2)
+       // Performance: Reuse tpe0, if possible.
+       if ((t1 eq tpe1) && (t2 eq tpe2)) {
+         tpe0
+       } else {
+         Type.Apply(t1, t2, loc)
+       }
+
+     case Type.Alias(sym, args, t, loc) =>
+       Type.Alias(sym, args.map(lowerType), lowerType(t), loc)
+
+     case Type.AssocType(cst, args, kind, loc) =>
+       // TODO: It appears that substitution (`reduceAssocType`) removes `AssocTypes`, so this should be an `InternalCompilerError`, right?
+       Type.AssocType(cst, args.map(lowerType), kind, loc)
+
+     case Type.JvmToType(_, loc) => throw InternalCompilerException("unexpected JVM type", loc)
+
+     case Type.JvmToEff(_, loc) => throw InternalCompilerException("unexpected JVM eff", loc)
+
+     case Type.UnresolvedJvmType(_, loc) => throw InternalCompilerException("unexpected JVM type", loc)
+
+   }
+
+  /**
+    * Returns the definition associated with the given symbol `sym`.
+    *
+    * @param tpe must be subst. Can be visited, if the underlying function can handle that
+    */
+  private def lookup(sym: Symbol.DefnSym, tpe: Type)(implicit ctx: Context, root: LoweredAst.Root, flix: Flix): Symbol.DefnSym =
+    Specialization.specializeDefnSym(sym, tpe)
```

Comment:
@CasperDN Maybe this is wrong. What if sym is already specialized? Then we duplicate work.

I think we need to ensure we use the cache.

Reply:
Hm.. Calling specializeDefSym does use the cache.